File: services/core-api/src/taleem_core/platform/ids.py
"""Identifier generation (pure-stdlib).

UUIDv7-style time-ordered ids: globally unique across contexts and offline clients, safe to
generate on an offline device, index-locality-friendly (docs/09-database-design.md §3).

Implemented without the (3.14-optional) uuid.uuid7 to keep it portable: 48-bit ms timestamp
prefix + 74 random bits, laid out per the RFC 9562 v7 field structure.
"""

from __future__ import annotations

import os
import time


def uuid7(now_ms: int | None = None) -> str:
    """Return a UUIDv7-shaped identifier as a canonical hex string.

    now_ms may be injected for determinism in tests. On real offline devices, prefer a
    server-synced clock where available; the random tail guarantees uniqueness regardless.
    """
    ts = now_ms if now_ms is not None else int(time.time() * 1000)
    ts &= (1 << 48) - 1
    rand = os.urandom(10)  # 80 bits; we use 74
    b = bytearray(16)
    b[0] = (ts >> 40) & 0xFF
    b[1] = (ts >> 32) & 0xFF
    b[2] = (ts >> 24) & 0xFF
    b[3] = (ts >> 16) & 0xFF
    b[4] = (ts >> 8) & 0xFF
    b[5] = ts & 0xFF
    b[6] = 0x70 | (rand[0] & 0x0F)  # version 7
    b[7] = rand[1]
    b[8] = 0x80 | (rand[2] & 0x3F)  # variant 10xx
    b[9:16] = rand[3:10]
    h = b.hex()
    return f"{h[0:8]}-{h[8:12]}-{h[12:16]}-{h[16:20]}-{h[20:32]}"

File: apps/web/lib/offline/ids.ts
// Client-side UUIDv7 (mirrors services/core-api platform/ids.py) — time-ordered, offline-safe.
// 48-bit ms timestamp prefix + 74 random bits, laid out per RFC 9562 v7.

export function uuid7(nowMs: number = Date.now(), rand: Uint8Array = randomBytes(10)): string {
  // 48-bit timestamp. Use division/modulo, not bitwise `&`/`>>>` — JS bit ops truncate to 32 bits
  // and would drop the top of a millisecond timestamp.
  const ts = nowMs % 2 ** 48;
  const b = new Uint8Array(16);
  b[0] = Math.floor(ts / 2 ** 40) % 256;
  b[1] = Math.floor(ts / 2 ** 32) % 256;
  b[2] = Math.floor(ts / 2 ** 24) % 256;
  b[3] = Math.floor(ts / 2 ** 16) % 256;
  b[4] = Math.floor(ts / 2 ** 8) % 256;
  b[5] = ts % 256;
  b[6] = 0x70 | ((rand[0] ?? 0) & 0x0f); // version 7
  b[7] = rand[1] ?? 0;
  b[8] = 0x80 | ((rand[2] ?? 0) & 0x3f); // variant 10xx
  b[9] = rand[3] ?? 0;
  b[10] = rand[4] ?? 0;
  b[11] = rand[5] ?? 0;
  b[12] = rand[6] ?? 0;
  b[13] = rand[7] ?? 0;
  b[14] = rand[8] ?? 0;
  b[15] = rand[9] ?? 0;
  const h = Array.from(b)
    .map((x) => x.toString(16).padStart(2, "0"))
    .join("");
  return `${h.slice(0, 8)}-${h.slice(8, 12)}-${h.slice(12, 16)}-${h.slice(16, 20)}-${h.slice(20, 32)}`;
}

function randomBytes(n: number): Uint8Array {
  const out = new Uint8Array(n);
  if (typeof globalThis.crypto?.getRandomValues === "function") {
    globalThis.crypto.getRandomValues(out);
  }
  return out;
}
